File: stations/agent-metrics/src/agent_metrics/routes/projects.py
"""Team-task project management routes."""

from __future__ import annotations

from fastapi import APIRouter, HTTPException

from agent_metrics.db import get_pool
from agent_metrics.engines import task_manager as tm
from agent_metrics.models import (
    DebaterAdd,
    ProjectCreate,
    RoundAction,
    TaskAdd,
    TaskResult,
    TaskUpdate,
)

router = APIRouter()


@router.get("/")
async def list_projects() -> list[dict]:
    pool = await get_pool()
    return await tm.list_projects(pool)


@router.post("/")
async def create_project(body: ProjectCreate) -> dict:
    pool = await get_pool()
    try:
        proj = await tm.init_project(
            pool,
            body.name,
            body.mode.value,
            goal=body.goal,
            pipeline=body.pipeline,
            workspace=body.workspace,
        )
        return {"status": "created", "project": proj}
    except FileExistsError as e:
        raise HTTPException(400, str(e)) from None
    except ValueError as e:
        raise HTTPException(422, str(e)) from None


@router.get("/{name}")
async def get_project(name: str) -> dict:
    pool = await get_pool()
    try:
        return await tm.get_status(pool, name)
    except FileNotFoundError as e:
        raise HTTPException(404, str(e)) from None


@router.post("/{name}/tasks")
async def add_task(name: str, body: TaskAdd) -> dict:
    pool = await get_pool()
    try:
        task = await tm.add_task(
            pool,
            name,
            body.task_id,
            agent=body.agent,
            desc=body.description,
            deps=body.deps,
        )
        return {"status": "added", "task": task}
    except (FileNotFoundError, ValueError) as e:
        raise HTTPException(400, str(e)) from None


@router.get("/{name}/ready")
async def ready_tasks(name: str) -> list[dict]:
    pool = await get_pool()
    try:
        return await tm.get_ready_tasks(pool, name)
    except (FileNotFoundError, ValueError) as e:
        raise HTTPException(400, str(e)) from None


@router.get("/{name}/next")
async def next_stage(name: str) -> dict:
    pool = await get_pool()
    try:
        stage = await tm.get_next_stage(pool, name)
        if stage is None:
            return {"status": "all_complete"}
        return stage
    except (FileNotFoundError, ValueError) as e:
        raise HTTPException(400, str(e)) from None


@router.patch("/{name}/tasks/{task_id}")
async def update_task(name: str, task_id: str, body: TaskUpdate) -> dict:
    pool = await get_pool()
    try:
        return await tm.update_task_status(pool, name, task_id, body.status.value)
    except (FileNotFoundError, ValueError) as e:
        raise HTTPException(400, str(e)) from None


@router.post("/{name}/tasks/{task_id}/result")
async def record_result(name: str, task_id: str, body: TaskResult) -> dict:
    pool = await get_pool()
    try:
        await tm.record_result(pool, name, task_id, body.text)
        return {"status": "recorded"}
    except (FileNotFoundError, ValueError) as e:
        raise HTTPException(400, str(e)) from None


@router.post("/{name}/debaters")
async def add_debater(name: str, body: DebaterAdd) -> dict:
    pool = await get_pool()
    try:
        debater = await tm.add_debater(
            pool,
            name,
            body.debater_id,
            agent=body.agent,
            perspective=body.perspective,
        )
        return {"status": "added", "debater": debater}
    except (FileNotFoundError, ValueError) as e:
        raise HTTPException(400, str(e)) from None


@router.post("/{name}/rounds")
async def manage_round(name: str, body: RoundAction) -> dict:
    pool = await get_pool()
    try:
        return await tm.manage_round(
            pool,
            name,
            body.action,
            debater_id=body.debater_id,
            text=body.text,
        )
    except (FileNotFoundError, ValueError) as e:
        raise HTTPException(400, str(e)) from None


@router.post("/{name}/reset")
async def reset_project(name: str) -> dict:
    pool = await get_pool()
    try:
        await tm.reset_project(pool, name)
        return {"status": "reset"}
    except FileNotFoundError as e:
        raise HTTPException(404, str(e)) from None

File: stations/agent-metrics-rs/src/web/projects.rs
//! /projects/* routes — team-task project CRUD + tasks + debaters + rounds.

use super::AppState;
use crate::engines::tasks::{self, TaskError};
use axum::{
    extract::{Path, State},
    http::StatusCode,
    Json,
};
use serde::Deserialize;
use serde_json::{json, Value};

fn map_err(e: TaskError) -> (StatusCode, Json<Value>) {
    let code = match &e {
        TaskError::NotFound(_) => StatusCode::NOT_FOUND,
        TaskError::Exists(_) => StatusCode::BAD_REQUEST,
        TaskError::Validation(_) => StatusCode::UNPROCESSABLE_ENTITY,
    };
    (code, Json(json!({"detail": e.to_string()})))
}

pub async fn list_projects(State(state): State<AppState>) -> Json<Value> {
    let rows = tasks::list_projects(&state.pool).await.unwrap_or_default();
    Json(serde_json::to_value(rows).unwrap_or(Value::Null))
}

#[derive(Debug, Deserialize)]
pub struct ProjectCreate {
    pub name: String,
    #[serde(default = "default_mode")]
    pub mode: String,
    #[serde(default)]
    pub goal: String,
    #[serde(default)]
    pub pipeline: String,
    #[serde(default)]
    pub workspace: String,
}
fn default_mode() -> String {
    "dag".into()
}

pub async fn create_project(
    State(state): State<AppState>,
    Json(body): Json<ProjectCreate>,
) -> Result<Json<Value>, (StatusCode, Json<Value>)> {
    let proj = tasks::init_project(
        &state.pool,
        &body.name,
        &body.mode,
        &body.goal,
        &body.pipeline,
        &body.workspace,
        false,
    )
    .await
    .map_err(map_err)?;
    Ok(Json(json!({"status": "created", "project": proj})))
}

pub async fn get_project(
    State(state): State<AppState>,
    Path(name): Path<String>,
) -> Result<Json<Value>, (StatusCode, Json<Value>)> {
    let v = tasks::get_status(&state.pool, &name).await.map_err(map_err)?;
    Ok(Json(v))
}

#[derive(Debug, Deserialize)]
pub struct TaskAdd {
    pub task_id: String,
    #[serde(default)]
    pub agent: String,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub deps: String,
}

pub async fn add_task(
    State(state): State<AppState>,
    Path(name): Path<String>,
    Json(body): Json<TaskAdd>,
) -> Result<Json<Value>, (StatusCode, Json<Value>)> {
    let t = tasks::add_task(&state.pool, &name, &body.task_id, &body.agent, &body.description, &body.deps)
        .await
        .map_err(map_err)?;
    Ok(Json(json!({"status": "added", "task": t})))
}

pub async fn ready_tasks(
    State(state): State<AppState>,
    Path(name): Path<String>,
) -> Result<Json<Value>, (StatusCode, Json<Value>)> {
    let v = tasks::get_ready_tasks(&state.pool, &name).await.map_err(map_err)?;
    Ok(Json(serde_json::to_value(v).unwrap_or(Value::Null)))
}

pub async fn next_stage(
    State(state): State<AppState>,
    Path(name): Path<String>,
) -> Result<Json<Value>, (StatusCode, Json<Value>)> {
    match tasks::get_next_stage(&state.pool, &name).await.map_err(map_err)? {
        Some(s) => Ok(Json(s)),
        None => Ok(Json(json!({"status": "all_complete"}))),
    }
}

#[derive(Debug, Deserialize)]
pub struct TaskUpdate {
    pub status: String,
}

pub async fn update_task(
    State(state): State<AppState>,
    Path((name, task_id)): Path<(String, String)>,
    Json(body): Json<TaskUpdate>,
) -> Result<Json<Value>, (StatusCode, Json<Value>)> {
    let v = tasks::update_task_status(&state.pool, &name, &task_id, &body.status)
        .await
        .map_err(map_err)?;
    Ok(Json(v))
}

#[derive(Debug, Deserialize)]
pub struct TaskResult {
    pub text: String,
}

pub async fn record_result(
    State(state): State<AppState>,
    Path((name, task_id)): Path<(String, String)>,
    Json(body): Json<TaskResult>,
) -> Result<Json<Value>, (StatusCode, Json<Value>)> {
    tasks::record_result(&state.pool, &name, &task_id, &body.text)
        .await
        .map_err(map_err)?;
    Ok(Json(json!({"status": "recorded"})))
}

#[derive(Debug, Deserialize)]
pub struct DebaterAdd {
    pub debater_id: String,
    #[serde(default)]
    pub agent: String,
    #[serde(default)]
    pub perspective: String,
}

pub async fn add_debater(
    State(state): State<AppState>,
    Path(name): Path<String>,
    Json(body): Json<DebaterAdd>,
) -> Result<Json<Value>, (StatusCode, Json<Value>)> {
    let d = tasks::add_debater(&state.pool, &name, &body.debater_id, &body.agent, &body.perspective)
        .await
        .map_err(map_err)?;
    Ok(Json(json!({"status": "added", "debater": d})))
}

#[derive(Debug, Deserialize)]
pub struct RoundAction {
    pub action: String,
    #[serde(default)]
    pub debater_id: String,
    #[serde(default)]
    pub text: String,
}

pub async fn manage_round(
    State(state): State<AppState>,
    Path(name): Path<String>,
    Json(body): Json<RoundAction>,
) -> Result<Json<Value>, (StatusCode, Json<Value>)> {
    let v = tasks::manage_round(&state.pool, &name, &body.action, &body.debater_id, &body.text)
        .await
        .map_err(map_err)?;
    Ok(Json(v))
}

pub async fn reset_project(
    State(state): State<AppState>,
    Path(name): Path<String>,
) -> Result<Json<Value>, (StatusCode, Json<Value>)> {
    tasks::reset_project(&state.pool, &name).await.map_err(map_err)?;
    Ok(Json(json!({"status": "reset"})))
}
